diff -auNrd git.orig/brcm/westeros-gl/westeros-gl.cpp git/brcm/westeros-gl/westeros-gl.cpp
--- git.orig/brcm/westeros-gl/westeros-gl.cpp	2019-08-16 18:57:50.089079532 -0400
+++ git/brcm/westeros-gl/westeros-gl.cpp	2019-08-16 18:55:55.900357876 -0400
@@ -25,6 +25,7 @@
 
 #include <EGL/egl.h>
 #include <EGL/eglext.h>
+#include <EGL/begl_displayplatform.h>
 
 #include "nexus_config.h"
 #include "nexus_platform.h"
